### ISSUE TITLE
Annotate unsupported APIs in System.Management

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/runtime#52087]
As part of https://github.com/dotnet/runtime/issues/47228 i am running an analyzer to detect APIs throwing PNSE but not annotated with Obsolete/SupportedOSPlatform/UnsupportedOSPlatform attributes, we need to annotate them so that developers get warnings when they use them unexpectedly

For now, I have results only for cross-platform builds, analysis of targeted builds are coming soon where more APIs could be detected

Location | API 
-- | -- 
libraries\System.Management\src\System\Management\ManagementScope.cs(389,38) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(390,60) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(391,73) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(392,79) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(393,100) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(394,67) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(395,63) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(396,39) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(397,71) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(398,116) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(399,95) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(400,40) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(401,109) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(402,50) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(403,121) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(404,43) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(405,92) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(406,61) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(407,79) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(408,83) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(409,79) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(410,49) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(411,86) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(412,64) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(413,108) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(414,64) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(415,46) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(416,56) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(417,113) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(418,49) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(419,90) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(420,84) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(421,108) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(422,92) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(423,49) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(424,85) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(425,59) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(426,113) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(427,52) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(428,80) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(429,77) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(430,210) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(431,207) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(432,221) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(433,233) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(434,111) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(435,108) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(436,168) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(437,191) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(438,36) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'
libraries\System.Management\src\System\Management\ManagementScope.cs(439,35) | 'WmiNetUtilsHelper.LoadPlatformNotSupportedDelegates(string)'

These are all found in the private method `LoadPlatformNotSupportedDelegates()` where all delegates throwing PNSE with a message that it needs NetFramemork and/or with a specific version. This means APIs using these delegates are might not supported on .Net Core and might need obsoletion.

Suggestion: Please check all references and if the referenced API is unconditionally throwing (running the delegate) then that API might need an Obsoletion. If the API is internal/private no need for obsoletion, but those internal/private methods might lead you to the public API needs obsoletion

Note: We are suggesting adding [Obsolete] for the APIs only supported in .Net framework but not supported in .NetCore, with the corresponding Message, DiagnosticId, and UrlFormat https://github.com/dotnet/runtime/blob/cecc76a29a8b6c1fca181a38f076b4d949081746/src/libraries/System.Net.Requests/src/System/Net/AuthenticationManager.cs#L17 You can claim the next available DiagnosticId from Obsoletions.cs .

cc @jeffhandley @terrajobst @GrabYourPitchforks

**Comment by deeprobin** [comment]
This is a api-change? Is this api-approved?

/cc @jeffhandley @terrajobst @GrabYourPitchforks

**Comment by ViktorHofer** [comment]
@buyaa-n @jeffhandley do you know if adding Supported/UnsupportedOSPlatform attributes to public API needs to be approved by .NET API Review board?

**Comment by buyaa-n** [comment]
These APIs are internal and it's not clear how these APIs should be annotated. This issue is about investigating why these APIs throw PNSE and if they need to be attributed and what attribute should be used (Obsoleted or SupportedOSPlatform or UnsupportedOSPlatform). 

Then the attribute should be applied which might cause warnings where the API is used, and call site also might need to be attributed (or needs platform check). Propagating the attribute to the call site might reach to public a API and that is the purpose of this investigation, we want to find public APIs that needs to be attributed. After finding the public APIs that needs to be attributed we can review them as needed

**Comment by buyaa-n** [comment]
Sounds like they could be Obsoleted as they not supported in .NET Core. The question is if they are affecting public APIs (if they unconditionally called in public API). If they are not affecting public APIs then no need to annotate.

**Comment by danmoseley** [comment]
> Sounds like they could be Obsoleted as they not supported in .NET Core.

They are fully supported in .NET Core (sorry, I wasn't clear). But for them to work in .NET Core, the .NET Framework must be installed. Essentially - they reach into .NET Framework's implementation in order to work. We did it this way because this is a legacy tech, applicable only to Windows anyway. and we did not want to bring over and maintain substantially more code.
